federation/pkg/federation-controller/service/servicecontroller.go

```diff
@@ -0,0 +1,830 @@
+	clone, err := conversion.NewCloner().DeepCopy(fedService)
+	if err != nil {
+		return err, !retryable
+	}
+	service, ok := clone.(*v1.Service)
+	if !ok {
+		return fmt.Errorf("Unexpected service cast error : %v\n", service), !retryable
+	}
+
+	// handle available clusters one by one
+	var hasErr bool
+	for clusterName, cluster := range s.clusterCache.clientMap {
+		go func(){
+			err = s.processServiceForCluster(clusterName, service, cluster.clientset)
+			if err != nil {
+				hasErr = true
+			}
+		}()
+	}
+	if hasErr {
+		// detail error has been dumpped inside the loop
+		return fmt.Errorf("service %s/%s was not successfully updated to all clusters", service.Namespace, service.Name), retryable
+	}
+	return nil, !retryable
+}
+
+func (s *ServiceController) deleteFederationService(service *v1.Service) (error, bool) {
+	// handle available clusters one by one
+	for clusterName, cluster := range s.clusterCache.clientMap {
+		s.deleteClusterService(clusterName, service, cluster.clientset)
+	}
+	return nil, !retryable
+}
+func (s *ServiceController) deleteClusterService(clusterName string, service *v1.Service, clientset *clientset.Clientset) (error, bool) {
+	glog.V(4).Infof("Deleting service %s/%s from cluster %s", service.Namespace, service.Name, clusterName)
+	var err error
+	for i := 0; i < clientRetryCount; i++ {
+		err = clientset.Core().Services(service.Namespace).Delete(service.Name, &api.DeleteOptions{})
+		if err == nil || errors.IsNotFound(err) {
+			glog.V(4).Infof("service %s/%s deleted from cluster %s", service.Namespace, service.Name, clusterName)
+			return nil, !retryable
+		}
+		time.Sleep(clientRetryInterval)
+	}
+	glog.V(4).Infof("failed to delete service %s/%s from cluster %s, %+v", service.Namespace, service.Name, clusterName, err)
+	return err, !retryable
+}
+
+func (s *ServiceController) ensureClusterServices(clusterName string, service *v1.Service, client *clientset.Clientset) error {
+	var err error
+	var needUpdate bool
+	for i := 0; i < clientRetryCount; i++ {
+		svc, err := client.Core().Services(service.Namespace).Get(service.Name)
+		if err == nil && svc != nil {
+			// service exists
+			glog.V(5).Infof("found service %s/%s from cluster %s", service.Namespace, service.Name, clusterName)
+			// TODO, what properties can be updated by service controller?
+			// possibily:
+			// metadata.labels
+			// spec.type spec.selector spec.ports
+			if !reflect.DeepEqual(svc.Labels, service.Labels) {
+				svc.Labels = service.Labels
+				needUpdate = true
+			}
+			if !reflect.DeepEqual(svc.Spec.Type, service.Spec.Type) {
+				svc.Spec.Type = service.Spec.Type
+				needUpdate = true
+			}
+			if !reflect.DeepEqual(svc.Spec.Selector, service.Spec.Selector) {
+				svc.Spec.Selector = service.Spec.Selector
+				needUpdate = true
+			}
+			// TODO: make sure auto assigned node port is updated to federation apiserver, otherwise value of fed service port will be 0
```

Comment:
This todo should be removed, as I've commented these lines
	// we do not need to check node port for federation
	//if x.NodePort != y.NodePort {
	//	return false
	//}